### PR TITLE
feat(autoware-ml): introduce metrics

### PR DESCRIPTION
**Pull request by amadeuszsz**
## Summary

Replaces the hand-rolled metric implementation with a composable framework built on `torchmetrics`.

**Core design**
- A per-task **suite** (`MetricSuite`, a `torchmetrics.Metric`) owns the accumulated state, its cross-GPU sync, and the per-range dispatch. It knows nothing about which metrics run.
- A **metric** (`Metric`) is a small, injectable, self-contained scorer: `evaluate(state, stage)` plus a configurable `stages`. The suite runs the metrics it was given (`components`) against the state it builds.
- `EvalStage` is `VAL` / `TEST`. Validation reports the light headline set, test the full set, gated per component via `stages`.

**Detection** (`autoware_ml/metrics/detection3d/`)
- `Detection3DMetricSuite` stores per-frame predictions/GT as list-states and applies the GT filters (range caps, min points, attribute exclusions) at `update`.
- Components: `MeanAP`, `HeadingAP`, `Nds`, `TpErrors`. Center-distance matching lives in `matching.py` and is memoized on `DetectionState` so it runs once per `(class, threshold)` and is shared across components.

**Segmentation** (`autoware_ml/metrics/segmentation3d/`)
- `Segmentation3DMetricSuite` keeps one stacked `(ranges+1, C, C)` confusion sum-state.
- Components: `IoU`, `Accuracy`, `PrecisionRecallF1`, derived from the shared `ConfusionState`. `pointwise.py` retains the step-level helpers.

**Range awareness**
- Every metric key is also emitted per radial `MetricRange` bucket with a distance suffix (e.g. `test/det3d/mAP_car_50m_90m`, `test/seg3d/iou_car_0m_50m`).

**Model integration**
- `MetricEvalMixin` drives the lifecycle (reset at epoch start, `update` per batch, `result` at epoch end, log under `{split}/{prefix}/{key}`). Models implement only `build_eval_output`.
- `model.metrics` is a list of suites; each suite lists its `components`. Removed `Detection3DBaseModel`; leaf models forward base args via `**kwargs`.
- New `seg_class_names` OmegaConf resolver builds per-index segmentation class names for metric keys (shared indices joined with `-`).

**CLI**
- `autoware-ml test` forces a single device by default for deterministic, padding-free evaluation. `--use-config-devices` honors `trainer.devices` from the config.

**Configs**
- Task configs inject the suite + components. Detection variants retune via the `metric_ranges` / `metric_eval_class_range` interpolation variables instead of restating the suite.

**Tests and docs**
- Per-component tests, full-suite value-parity tests, multi-rank gloo DDP sync tests (world size 2 and 8), and stage-gating tests.
- `docs/framework/metrics.md` documents the design.

## Checklist

- [ ] You've checked our [contribution overview](https://tier4.github.io/autoware-ml/contributing/contribution-overview/).
- [ ] Your PR follows our [pull request guidelines](https://tier4.github.io/autoware-ml/contributing/pull-request-guidelines/).

## Additional Notes

**Behavior preserving**
- Metric keys and values are unchanged from the previous implementation; the existing value tests act as the safety net.

**Distributed correctness**
- Sync is handled by `add_state`: detection per-frame tensors use list-states (`dist_reduce_fx=None`, which preserves per-frame grouping after the gather), segmentation confusion uses `sum`. Logged without `sync_dist` since state is identical on every rank after sync.
- Single-device test default avoids DistributedSampler padding (double-counts at most `world_size - 1` frames). Multi-device eval still works via `--use-config-devices`, with the padding caveat documented in `metrics.md`.

**Design choices**
- `model.metrics` is a list, and Hydra replaces a list wholesale, so detection variants override the two interpolation vars rather than restating the whole suite.
- Suites are generic over their state type (`MetricSuite[DetectionState]`); this is type-checker-only and erased at runtime.
- Components own their computation. Some redundancy is accepted on purpose (for example NDS recomputes matching) in exchange for self-contained, injectable metrics; the per-state matching cache removes the redundancy within a bucket.

**Out of scope (land with their own PRs)**
- Truck+trailer merge config hunks, transhead pooling, the visualization CLI/tests, and the new detection model families. Their configs reference the new suite but are not part of this PR.

### CODE REVIEW

File: autoware_ml/metrics/detection3d/eval_output.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+"""Shared eval-output builder for 3D detection models.
+
+Every detection model decodes its head into per-sample predictions and pairs
+them with the ground-truth boxes and labels. This helper builds the flat
+eval-output dict that :class:`~autoware_ml.metrics.detection3d.suite.AutowareDetection3DMetrics`
+reads, so each model's ``build_eval_output`` is a one-line delegation.
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/metrics/detection3d/eval_output.py:6]: https://github.com/tier4/autoware-ml/pull/49/commits/b65cfbc0a659019e296a0e5f7d699ef5a6ebf443

File: autoware_ml/metrics/detection3d/matching.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    MatchCurve,
+    PredictionRecord,
+    SelectedTpErrors,
+)
+
+
+def to_cpu_tensor(tensor: torch.Tensor, dtype: torch.dtype | None = None) -> torch.Tensor:
+    """Move a tensor to CPU and optionally cast its dtype."""
+    output = tensor.detach().to(device="cpu")
+    return output.to(dtype=dtype) if dtype is not None else output
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/metrics/detection3d/matching.py]: https://github.com/tier4/autoware-ml/pull/49/commits/e3c7396eef93f04b8772cc3fa1abf57dcca94ea8

File: autoware_ml/metrics/detection3d/matching.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+def _max_f1_index(f1_scores: np.ndarray) -> int:
+    if f1_scores.shape[0] == 0 or np.all(np.isnan(f1_scores)):
+        return -1
+    return int(np.nanargmax(f1_scores))
+
+
+def select_recall_tp_errors(curve: MatchCurve, recall_target: float) -> SelectedTpErrors:
+    """Mean TP errors over the matches up to a recall target."""
+    effective_recall = (int(round(100 * recall_target)) + 1) / 100.0
+    target_matches = int(np.floor(curve.total_gt * effective_recall))
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/metrics/detection3d/matching.py:317]: https://github.com/tier4/autoware-ml/pull/49/commits/42b1c81a466f7b3fb37c46f550fb32a03a9f9376

File: autoware_ml/metrics/detection3d/nds.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        error_dicts: list[dict[str, float]] = []
+        for label in labels:
+            aps: list[float] = []
+            aphs: list[float] = []
+            for threshold in state.thresholds:
+                curve = state.match_curve(label, threshold)
+                metrics = curve_metrics(curve)
+                aps.append(metrics.ap)
+                aphs.append(metrics.aph)
+                error_dicts.append(select_recall_tp_errors(curve, self.recall_target).errors)
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/metrics/detection3d/nds.py]: https://github.com/tier4/autoware-ml/pull/49/commits/42b1c81a466f7b3fb37c46f550fb32a03a9f9376

File: autoware_ml/metrics/detection3d/tp_errors.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        variants: dict[str, dict[tuple[int, float], SelectedTpErrors]] = {
+            name: {} for name in self.recall_targets
+        }
+        variants["optimal"] = {}
+
+        for label in labels:
+            for threshold in state.thresholds:
+                curve = state.match_curve(label, threshold)
+                for name, target in self.recall_targets.items():
+                    variants[name][(label, threshold)] = select_recall_tp_errors(curve, target)
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/metrics/detection3d/tp_errors.py:49]: https://github.com/tier4/autoware-ml/pull/49/commits/42b1c81a466f7b3fb37c46f550fb32a03a9f9376
